File: src/base_controller/scripts/base_controller.py
#!/usr/bin/env python
import rospy
from std_msgs.msg import Float32MultiArray
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Point,Pose,Quaternion,Twist,Vector3
import numpy as np
from math import sin,cos,pi
import tf


class baseController:
    def __init__(self):
        self.Xk = np.zeros((5,1),dtype=np.float64)
        self.Pk = np.eye(5,dtype=np.float64)
        self.Q = np.eye(5,dtype=np.float64)
        self.Q[0,0] = 0.25
        self.Q[1,1] = 0.25
        self.Q[2,2] = 0.27
        self.Q[3,3] = 0.01
        self.Q[4,4] = 0.1

        self.H1 = np.array([[0,0,0,1,0],[0,0,0,0,1]],dtype=np.float64)
        self.R1 = np.array([[0.01,0],[0,0.1]],dtype=np.float64)

        self.H2 = np.array([[0,0,0,0,1]],dtype=np.float64)
        self.R2 = np.array([[0.1]],dtype=np.float64)

        self.G = np.eye(5,dtype=np.float64)


        rospy.init_node('odometry_publisher')
        self.pub = rospy.Publisher('odom', Odometry, queue_size=10)
        self.tf_pub = tf.TransformBroadcaster()
        rospy.Subscriber("base_controller_data", Float32MultiArray, self.callback)
        rospy.spin()

    def callback(self,msg):
        dt = msg.data[8]
        self.G[0,2] = -self.Xk[3,0]*sin(self.Xk[2,0])*dt
        self.G[0,3] = dt*cos(self.Xk[2,0])
        self.G[1,2] = self.Xk[3,0]*cos(self.Xk[2,0])*dt
        self.G[1,3] = dt*sin(self.Xk[2,0])
        self.G[2,4] = dt

        self.Xk += np.array([[self.Xk[3,0]*cos(self.Xk[2,0])*dt],[self.Xk[3,0]*sin(self.Xk[2,0])*dt],[self.Xk[4,0]*dt],[0],[0]])
        self.Pk = self.G.dot(self.Pk).dot(self.G.T) + self.Q

        vx = msg.data[3]
        vth = msg.data[4]

        g_factor = 1.125
        Z1 = np.array([[vx],[vth*g_factor]],dtype=np.float64)

        S = self.H1.dot(self.Pk).dot(self.H1.T) + self.R1

        K = self.Pk.dot(self.H1.T).dot(np.linalg.pinv(S))

        self.Xk += K.dot(Z1 - self.H1.dot(self.Xk))

        self.Pk -= K.dot(self.H1).dot(self.Pk)

        gyro_rate = msg.data[6]*pi/180.0

        Z2 = np.array([[gyro_rate*g_factor]],dtype=np.float64)

        S = self.H2.dot(self.Pk).dot(self.H2.T) + self.R2

        K = self.Pk.dot(self.H2.T)/S

        # rospy.loginfo(str(K.shape))
        # rospy.loginfo(str(S.shape))
        # rospy.loginfo(str(Z2.shape))
        # rospy.loginfo(str(self.H2.shape))
        self.Xk += K.dot(Z2 - self.H2.dot(self.Xk))

        self.Pk -= K.dot(self.H2).dot(self.Pk)

        odom_quat = tf.transformations.quaternion_from_euler(0,0,self.Xk[2,0])
        current_time = rospy.Time.now()

        self.tf_pub.sendTransform((self.Xk[0,0],self.Xk[1,0],0),odom_quat,current_time,"base_link","odom")


        odom_msg = Odometry()

        odom_msg.header.stamp = current_time
        odom_msg.header.frame_id = "odom"

        
        odom_msg.pose.pose = Pose(Point(self.Xk[0,0],self.Xk[1,0],0),Quaternion(*odom_quat))
        odom_msg.pose.covariance = self.setPoseCovariance()


        odom_msg.child_frame_id = "base_link"
        odom_msg.twist.twist = Twist(Vector3(self.Xk[3,0],0,0),Vector3(0,0,self.Xk[4,0]))
        odom_msg.twist.covariance = self.setTwistCovariance()

        self.pub.publish(odom_msg)

    def setPoseCovariance(self):
        pose_covariance = [0]*36

        pose_covariance[0] = self.Pk[0,0]
        pose_covariance[1] = self.Pk[0,1]
        pose_covariance[5] = self.Pk[0,2]
        pose_covariance[6] = self.Pk[1,0]
        pose_covariance[7] = self.Pk[1,1]
        pose_covariance[11] = self.Pk[1,2]
        pose_covariance[30] = self.Pk[2,0]
        pose_covariance[31] = self.Pk[2,1]
        pose_covariance[35] = self.Pk[2,2]

        return pose_covariance 

    def setTwistCovariance(self):
        twist_covariance = [0]*36

        twist_covariance[0] = self.Pk[3,3]
        twist_covariance[5] = self.Pk[3,4]
        twist_covariance[30] = self.Pk[4,3]
        twist_covariance[35] = self.Pk[4,4]
        
        return twist_covariance
        

if __name__ == '__main__':
    baseController()

File: src/base_controller/src/base_controller.cpp
#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <nav_msgs/Odometry.h>
#include <std_msgs/Float32MultiArray.h>
#include <std_msgs/Bool.h>
#include <sensor_msgs/Imu.h>
#include <std_srvs/SetBool.h>

//#define KINECT_BASE_ANGLE 0.0//-0.65
#define M_PI           3.14159265358979323846
//#define GYRO_OFFSET 0.0125

class baseController{
  public:
    baseController(){
      ros::NodeHandle nh_;

      ros::NodeHandle nhLocal_("~");
      nhLocal_.getParam("gyro_offset",gyro_offset);
      nhLocal_.getParam("tf_data_teleop", tf_data_teleop);
      nhLocal_.getParam("tf_data_human", tf_data_human);

      sub_ = nh_.subscribe("base_controller_data", 1000, &baseController::baseCallback,this);
      pub_ = nh_.advertise<nav_msgs::Odometry>("wheel_odom", 50);
      imuPub_ = nh_.advertise<sensor_msgs::Imu>("imu_data", 50);
      

      service_ = nh_.advertiseService("robot_mode_change", &baseController::serviceCallback,this);
      pubMode_ = nh_.advertise<std_msgs::Bool>("robot_mode", 50);

      // ROS_ERROR("%f",gyro_offset);

    } 

    void baseCallback(const std_msgs::Float32MultiArray::ConstPtr& msg)
    {
      double x = msg->data[0];
      double y = msg->data[1];
      double th = msg->data[2];
      double vx = msg->data[3];
      double vth = msg->data[4];
	    double pitch = msg->data[5];
      double yaw_rate = msg->data[6];
      double accelerometer_rate = msg->data[7];
      teleop_mode = msg->data[8];

      // ROS_ERROR("%d",teleop_mode);

      ros::Time current_time = ros::Time::now();

	  //---------------imu_publish-----------------------------------------
      yaw_rate = yaw_rate/180.0*M_PI - gyro_offset;
      sensor_msgs::Imu imu_msg;
      imu_msg.header.stamp = current_time;
      imu_msg.header.frame_id = "imu_link";

      imu_msg.angular_velocity.x = 0;
      imu_msg.angular_velocity.y = 0;
      imu_msg.angular_velocity.z = yaw_rate;
      imu_msg.angular_velocity_covariance = { 0, 0, 0, 0, 0, 0, 0, 0, 0.001};

      imu_msg.linear_acceleration.x = accelerometer_rate;
      imu_msg.linear_acceleration.y = 0; 
      imu_msg.linear_acceleration.z = 9.8;
      imu_msg.linear_acceleration_covariance = { 0.01, 0, 0, 0, 0, 0, 0, 0, 0 };

      imu_msg.orientation_covariance = {-1,   0,   0, 0,  -1,   0, 0,   0,  -1 };

      imuPub_.publish(imu_msg);
     
	  //----------------------------------------------------------------------
	  
	  //-------------base_link_to_pitch_link---------------------------------
      geometry_msgs::TransformStamped odom_trans;
      odom_trans.header.stamp = current_time;
      odom_trans.header.frame_id = "base_link";
      odom_trans.child_frame_id = "pitch_link";

      odom_trans.transform.translation.x = 0.0;
      odom_trans.transform.translation.y = 0.0;
      odom_trans.transform.translation.z = 0.0;
      odom_trans.transform.rotation =  tf::createQuaternionMsgFromRollPitchYaw(0,(pitch)*M_PI/180.0,0);

      //send the transform
      tf_broadcaster_.sendTransform(odom_trans);

      //-------------------------------------------------------------------

      //---------static_tfs------------------------------------------------
      if(teleop_mode == 1){

        geometry_msgs::TransformStamped static_trans;
        static_trans.header.stamp = current_time;
        static_trans.header.frame_id = "base_link";
        static_trans.child_frame_id = "lidar_link";

        static_trans.transform.translation.x = tf_data_teleop[0];
        static_trans.transform.translation.y = 0.0;
        static_trans.transform.translation.z = tf_data_teleop[1];
        static_trans.transform.rotation =  tf::createQuaternionMsgFromRollPitchYaw(0,0,0);

        tf_broadcaster_.sendTransform(static_trans);

        static_trans.header.stamp = current_time;
        static_trans.header.frame_id = "pitch_link";
        static_trans.child_frame_id = "camera_link";

        static_trans.transform.translation.x = tf_data_teleop[2];
        static_trans.transform.translation.y = 0.0;
        static_trans.transform.translation.z = tf_data_teleop[3];
        static_trans.transform.rotation =  tf::createQuaternionMsgFromRollPitchYaw(0,0.17,0);

        tf_broadcaster_.sendTransform(static_trans);

      } else if (teleop_mode == 0){

        geometry_msgs::TransformStamped static_trans;
        static_trans.header.stamp = current_time;
        static_trans.header.frame_id = "base_link";
        static_trans.child_frame_id = "lidar_link";

        static_trans.transform.translation.x = tf_data_human[0];
        static_trans.transform.translation.y = 0.0;
        static_trans.transform.translation.z = tf_data_human[1];
        static_trans.transform.rotation =  tf::createQuaternionMsgFromRollPitchYaw(0,0,0);

        tf_broadcaster_.sendTransform(static_trans);

        static_trans.header.stamp = current_time;
        static_trans.header.frame_id = "pitch_link";
        static_trans.child_frame_id = "camera_link";

        static_trans.transform.translation.x = tf_data_human[2];
        static_trans.transform.translation.y = 0.0;
        static_trans.transform.translation.z = tf_data_human[3];
        static_trans.transform.rotation =  tf::createQuaternionMsgFromRollPitchYaw(0,0.17,0);

        tf_broadcaster_.sendTransform(static_trans);

      }


      //------------------------------------------------------------------


	  //-----------------------------------------------------------------------

      nav_msgs::Odometry odom;
      odom.header.stamp = current_time;
      odom.header.frame_id = "odom";

      //set the position
      odom.pose.pose.position.x = x;
      odom.pose.pose.position.y = y;
      odom.pose.pose.position.z = 0.0;
      odom.pose.pose.orientation = tf::createQuaternionMsgFromYaw(th);
	    odom.pose.covariance[0] = 0.25;
	    odom.pose.covariance[7] = 0.25;
      odom.pose.covariance[14] = 0.25;
	    odom.pose.covariance[21] = 0.1;
	    odom.pose.covariance[28] = 0.1;
      odom.pose.covariance[35] = 0.1;

      //set the velocity
      odom.child_frame_id = "base_link";
      odom.twist.twist.linear.x = vx;
      odom.twist.twist.linear.y = 0;
      odom.twist.twist.angular.z = vth;
  	  odom.twist.covariance[0] = 0.01;
  	  odom.twist.covariance[7] = 0.01;
  	  odom.twist.covariance[14] = 0.01;
  	  odom.twist.covariance[21] = 0.01;
  	  odom.twist.covariance[28] = 0.01;
  	  odom.twist.covariance[35] = 0.02;


      //publish the message
      pub_.publish(odom);

    }

    bool serviceCallback(std_srvs::SetBool::Request& request, std_srvs::SetBool::Response& response)
      {
        std_msgs::Bool msg;

        msg.data = request.data;

        pubMode_.publish(msg);

        response.success = true;

        if (msg.data == true){
          response.message = "mode changed to teleop";
        }else{
           response.message = "mode changed to human control";
        }

        return true;
      }

  protected:
    ros::Subscriber sub_;
    ros::Publisher pub_;
    ros::Publisher imuPub_;
    tf::TransformBroadcaster tf_broadcaster_;

    ros::ServiceServer service_;
    ros::Publisher pubMode_;

    double gyro_offset;
    bool teleop_mode;
    std::vector<double> tf_data_teleop; //teleop - (x1,z1) -- lidar,(x2,z2) -- kinect
    std::vector<double> tf_data_human; // human - (x3,z3),(x4,z4)

};



int main(int argc, char** argv){
  ros::init(argc, argv, "odometry_publisher");

  baseController base_controller;

  ros::spin();
  
}
